core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala

```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
         topKs = $(topKs)
       )
 
+      val rddCm = dataUse.select(col(labelColName), col(predictionColName), col(probabilityColName)).rdd.map{
+        case Row(label: Double, pred: Double, prob: DenseVector) => (label, pred, prob.toArray)
+      }
+      val confusionMatrixByThreshold = calculateConfMatrixMetricsByThreshold(rddCm)
+      val misClassifications = calculateMisClassificationMetrics( rddCm.map{ case (label, pred, _) => (label, pred)} )
+
       val metrics = MultiClassificationMetrics(
         Precision = precision,
         Recall = recall,
         F1 = f1,
         Error = error,
         ThresholdMetrics = thresholdMetrics,
-        TopKMetrics = topKMetrics
+        TopKMetrics = topKMetrics,
+        ConfusionMatrixMetrics = confusionMatrixByThreshold,
+        MisClassificationMetrics = misClassifications
       )
 
       log.info("Evaluated metrics: {}", metrics.toString)
       metrics
     }
   }
 
+/**
+ * function to construct the confusion matrix for the top n most occurring labels
+ * @param labelPredictionCtRDD RDD of ((label, prediction, confidence), count)
+ * @param cmClasses the top n most occurring classes, sorted by counts in descending order
+ * @return an array of counts
+ */
+  def constructConfusionMatrix(
+    labelPredictionCtRDD: RDD[((Double, Double, Double), Long)],
```

Comment:
Why not define a `case class LabelPredConf(label: Double, prediction: confidence: Double, count: Long)`?

Reply:
@tovbinm this case class is only used within this class for confusion matrix calculation and not sent outbound, so I didn't explicitly define it. In a future PR we can use Dataframe APIs/Row object instead of RDD tuples (suggestion from @gerashegalov ) in OpMultiClassificationEvaluator.scala, so that it's more explicit than the tuples and we don't have to declare it every time in the anonymous functions

Reply:
this can be a private case class for this module.

Reply:
If defining a case class `LabelPredConf(label: Double, prediction: confidence: Double, count: Long)`, it would require an additional map function to map the key-val pairs to this case class after the ```    val labelPredictionConfidenceCountRDD = dataTopNLabels.map{
      case (label, prediction, proba) => {
        ( (label, prediction, SearchHelper.findThreshold(sortedThresholds, proba.max)), 1L )
      }
    }.reduceByKey(_ + _) ```

And in then in `constructConfusionMatrix`, it will be `    val confusionMatrixMap = labelPredictionCtRDD.map( e => ((e.Label, e.Prediction), e.Count)).reduceByKey(_ + _).collectAsMap() `

Reply:
It's not mandatory. If it only scoped within one function / file - no need to create a case class.

Reply:
If defining a case class here, the new constructConfusionMatrix would look like 

 ```
 def constructConfusionMatrix(
    labelPredictionCtRDD: RDD[LabelPredictionConfidenceCt],
    cmClasses: Seq[Double]): Seq[Long] = {

    val confusionMatrixMap = labelPredictionCtRDD.map( r => ((r.Label, r.Prediction), r.Count))
      .reduceByKey(_ + _).collectAsMap()
```
And an additional map function will be needed  after `reduceByKey` in 

```
    val labelPredictionConfidenceCountRDD = dataTopNLabels.map{
      case (label, prediction, proba) => {
        ( (label, prediction, SearchHelper.findThreshold(sortedThresholds, proba.max)), 1L )
      }
    }.reduceByKey(_ + _)
      .map {
        case ((label, pred, confidence), ct) => LabelPredictionConfidenceCt(label, pred, confidence, ct)
      }
```
This case class is only used twice in the entire class. It doesn't seem to make the code more succinct. What would be the benefit of doing it? @tovbinm